<a href="https://colab.research.google.com/github/rio-gunawan18/TugasProyek_026_335/blob/main/Recomendation_song_spotify.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Rekomendasi Lagu pada Spotify**

Spotify merupakan layanan musik digital, podcast, dan video yang memberikan akses ke jutaan lagu dan konten lain dari kreator di seluruh dunia. Dari dataset ini bisa menjadi rekomendasi bagi penikmat musik yang ingin mendengarkan musik pada aplikasi spotify.

Analisis yang dilakukan ini untuk menjawab beberapa pertanyaan terkait

Problem Question :

1.   Bagaimana distribusi rata-rata popularitas lagu di setiap genre? Genre mana yang memiliki rata-rata popularitas tertinggi?
2.   Siapa penyanyi dengan jumlah lagu terbanyak?
3.   Berapa jumlah lagu yang diciptakan pada setiap genre?
4.   Apakah penyanyi tertentu cenderung memiliki popularitas lagu yang lebih tinggi dibandingkan penyanyi lainnya?
5. Apakah ada pola tertentu yang menunjukkan hubungan antara tahun rilis album dan popularitas lagunya?

Hasil analisis ini dapat memberikan wawasan yang strategis kepada produser musik dan artis:

1. Dapat memilih genre yang lebih disukai oleh pendengar dan menyesuaikan produksi musik dengan tren yang ada, sehingga meningkatkan peluang kesuksesan.
2. Merancang strategi rilis album yang lebih tepat waktu berdasarkan analisis popularitas lagu dan tahun rilis, mengoptimalkan dampak pemasaran dan distribusi.
3. Memaksimalkan promosi dan kolaborasi dengan artis yang memiliki tingkat popularitas tertinggi, serta memperluas jangkauan audiens melalui kolaborasi strategis.
4. Menyesuaikan dengan preferensi audiens yang lebih besar untuk memilih genre atau artis yang lebih banyak diminati, meningkatkan potensi keuntungan dan loyalitas penggemar.









**Dari dataset ini, kami memilih variabel-variabel berikut untuk dianalisis:**
1. playlist_genre
2. track_popularity
3. track_name
4. playlist_genre
5. track_artist
6. track_album_release_date

**Metodologi yang Digunakan :**
1. Menangani Missing Value
2. Exploratory Data Analysis (EDA) yaitu mengidentifikasi populer dari suatu lagi berdasarkan genre , dan penyanyi
3. Analisis Waktu populer suatu lagu yaitu menggunakan variabel track_album_release_date, dan track_popularity
4. Menyusun wawasan dari analisis yang dilakukan.



Data yang digunakan dalam analisis ini berasal dari github ([Github](https://github.com/rfordatascience/tidytuesday/blob/main/data/2020/2020-01-21/spotify_songs.csv)).

Data ini berasal dari Spotify dan diakses melalui package spotify, yang memudahkan pengambilan data pribadi atau metadata lagu dari API Spotify. Kaylin Pavlik menggunakan package ini untuk mengumpulkan 5000 lagu dari 6 genre utama (EDM, Latin, Pop, R&B, Rap, Rock) dan menganalisis fitur audionya untuk eksplorasi serta klasifikasi lagu.


Berikut penjelasan variable dari dataset :

| Atribut                       | Keterangan                                                                                       |
|-------------------------------|-------------------------------------------------------------------------------------------------|
| track_id| Song unique ID                             |
| track_name                  | Nama Lagu                                      |
| track_artist                   | Penyanyi dari lagu.        |
| track_popularity             | Popularitas Lagu dengan indikator (0-100) semakin tinggi nilai nya semakin populer                                                           |
| track_album_id          | Album unique ID                                                          |
| track_album_name	      | Nama album dari lagu                                     |
|track_album_release_date   | Tanggal Rilis Album                         |
| playlist_name       | Nama Playlist|
| playlist_id         | Playlist ID |
| playlist_genre                   | Playlist Genre                                                           |
| playlist_subgenre                 | Playlist subgenre                                                             |
| danceability                       | menggambarkan seberapa cocok suatu lagu untuk menari berdasarkan kombinasi elemen musik termasuk tempo, stabilitas ritme, kekuatan ketukan, dan keteraturan secara keseluruhan.                                                                   |
| Energy                          | Energi adalah ukuran dari 0,0 hingga 1,0 dan mewakili ukuran persepsi intensitas dan aktivitas|
|     key       | Perkiraan kunci keseluruhan trek. Integer dipetakan ke nada menggunakan notasi Kelas Pitch standar. Misalnya. 0 = C, 1 = C♯/D♭, 2 = D, dan seterusnya. Jika tidak ada kunci yang terdeteksi, nilainya -1                                     |
| loudness              | Kenyaringan keseluruhan trek dalam desibel (dB). Nilai kenyaringan dirata-ratakan di seluruh trek dan berguna untuk membandingkan kenyaringan relatif trek         |
| mode        | Mode menunjukkan modalitas (mayor atau minor) suatu lagu, jenis tangga nada yang menjadi asal isi melodinya. Mayor diwakili oleh 1 dan minor adalah 0.       |
| speechiness          | Speechiness mendeteksi keberadaan kata-kata yang diucapkan dalam sebuah trek.                            |
| acousticness     | Ukuran keyakinan dari 0,0 hingga 1,0 untuk menentukan apakah trek bersifat akustik atau tidak. 1.0 mewakili keyakinan tinggi bahwa trek tersebut bersifat akustik.                                         |
| instrumentalness| Memprediksi apakah suatu lagu tidak mengandung vokal. Bunyi "Ooh" dan "aah" dianggap instrumental dalam konteks ini. Lagu rap atau kata-kata yang diucapkan jelas bersifat "vokal".                                  |
| liveness | Mendeteksi kehadiran penonton dalam rekaman.                                  |
| valence| Ukuran dari 0,0 hingga 1,0 yang menggambarkan kepositifan musik yang disampaikan oleh sebuah lagu. |
| tempo| Perkiraan tempo keseluruhan trek dalam ketukan per menit (BPM). |
| duration_ms| Durasi lagu dalam milidetik |



Tujuan awal data yaitu untuk menyediakan informasi tentang lagu yang berada dalam platform spotify, mulai dari jenis musik , tipe musik , isi musik





In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from plotly.offline import init_notebook_mode
import seaborn as sns
import datetime as dt
import warnings

In [ ]:
url = "https://drive.google.com/uc?id=1uah4vbsvrGHsjOV8bCGgZpzgQVRrhCdk"
df = pd.read_csv(url)
df

track_id  \
0      6f807x0ima9a1j3VPbc7VN   
1      0r7CVbZTWZgbTCYdfa2P31   
2      1z1Hg7Vb0AhHDiEmnDE79l   
3      75FpbthrwQmzHlBJLuGdC7   
4      1e8PAfcKUYoKkxPhrHqw4x   
...                       ...   
32828  7bxnKAamR3snQ1VGLuVfC1   
32829  5Aevni09Em4575077nkWHz   
32830  7ImMqPP3Q1yfUHvsdn7wEo   
32831  2m69mhnfQ1Oq6lGtXuYhgX   
32832  29zWqhca3zt5NsckZqDf6c   

                                              track_name      track_artist  \
0      I Don't Care (with Justin Bieber) - Loud Luxur...        Ed Sheeran   
1                        Memories - Dillon Francis Remix          Maroon 5   
2                        All the Time - Don Diablo Remix      Zara Larsson   
3                      Call You Mine - Keanu Silva Remix  The Chainsmokers   
4                Someone You Loved - Future Humans Remix     Lewis Capaldi   
...                                                  ...               ...   
32828               City Of Lights - Official Radio Edit      Lush & Simon   
32829                Closer - Sultan & Ned Shepard Remix    Tegan and Sara   
32830                       Sweet Surrender - Radio Edit       Starkillers   
32831                     Only For You - Maor Levi Remix            Mat Zo   
32832                             Typhoon - Original Mix      Julian Calor   

       track_popularity          track_album_id  \
0                    66  2oCs0DGTsRO98Gh5ZSl2Cx   
1                    67  63rPSO264uRjW1X5E6cWv6   
2                    70  1HoSmj2eLcsrR0vE9gThr4   
3                    60  1nqYsOef1yKKuGOVchbsk6   
4                    69  7m7vv9wlQ4i0LFuJiE2zsQ   
...                 ...                     ...   
32828                42  2azRoBBWEEEYhqV6sb7JrT   
32829                20  6kD6KLxj7s8eCE3ABvAyf5   
32830                14  0ltWNSY9JgxoIZO4VzuCa6   
32831                15  1fGrOkHnHJcStl14zNx8Jy   
32832                27  0X3mUOm6MhxR7PzxG95rAo   

                                        track_album_name  \
0      I Don't Care (with Justin Bieber) [Loud Luxury...   
1                        Memories (Dillon Francis Remix)   
2                        All the Time (Don Diablo Remix)   
3                            Call You Mine - The Remixes   
4                Someone You Loved (Future Humans Remix)   
...                                                  ...   
32828                         City Of Lights (Vocal Mix)   
32829                                     Closer Remixed   
32830                       Sweet Surrender (Radio Edit)   
32831                             Only For You (Remixes)   
32832                                      Typhoon/Storm   

      track_album_release_date    playlist_name             playlist_id  \
0                   2019-06-14        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
1                   2019-12-13        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
2                   2019-07-05        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
3                   2019-07-19        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
4                   2019-03-05        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
...                        ...              ...                     ...   
32828               2014-04-28  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32829               2013-03-08  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32830               2014-04-21  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32831               2014-01-01  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32832               2014-03-03  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   

      playlist_genre  ... key  loudness  mode  speechiness  acousticness  \
0                pop  ...   6    -2.634     1       0.0583      0.102000   
1                pop  ...  11    -4.969     1       0.0373      0.072400   
2                pop  ...   1    -3.432     0       0.0742      0.079400   
3                pop  ...   7    -3.778     1       0.1020      0.028700   
4                pop  ...   1    -4.672     1       0.0359      0.080300

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   track_id                  32833 non-null  object 
 1   track_name                32828 non-null  object 
 2   track_artist              32828 non-null  object 
 3   track_popularity          32833 non-null  int64  
 4   track_album_id            32833 non-null  object 
 5   track_album_name          32828 non-null  object 
 6   track_album_release_date  32833 non-null  object 
 7   playlist_name             32833 non-null  object 
 8   playlist_id               32833 non-null  object 
 9   playlist_genre            32833 non-null  object 
 10  playlist_subgenre         32833 non-null  object 
 11  danceability              32833 non-null  float64
 12  energy                    32833 non-null  float64
 13  key                       32833 non-null  int64  
 14  loudne

# Check Missing Value

In [ ]:
df.isnull().sum()

track_id                    0
track_name                  5
track_artist                5
track_popularity            0
track_album_id              0
track_album_name            5
track_album_release_date    0
playlist_name               0
playlist_id                 0
playlist_genre              0
playlist_subgenre           0
danceability                0
energy                      0
key                         0
loudness                    0
mode                        0
speechiness                 0
acousticness                0
instrumentalness            0
liveness                    0
valence                     0
tempo                       0
duration_ms                 0
dtype: int64

Terdapat Missing Value pada Track_Album_Name,track_name,track_artist


*   Kolom `track_name`, `track_artist`, `track_album_name`  memiliki missing value karena mungkin terdapat lupa input pada kolom tersebut



In [ ]:
# prompt: hapus missing value

# Drop rows with missing values in 'track_album_name'
df = df.dropna(subset=['track_album_name'])

# Verify that the missing values have been removed
print(df.isnull().sum())

# Checking Duplicate

In [ ]:
df.duplicated().value_counts()

False    32833
Name: count, dtype: int64

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.shape

(32833, 23)

In [ ]:
# prompt: buatkan describe untuk column bertipe object

# melihat deskripsi kolom dengan tipe data object
for col in df.select_dtypes(include=['object']).columns:
    print(f"Column: {col}")
    print(df[col].describe())
    print("-" * 20)

Column: track_id
count                      32833
unique                     28356
top       7BKLCZ1jbUBVqRi2FVlTVw
freq                          10
Name: track_id, dtype: object
--------------------
Column: track_name
count      32828
unique     23449
top       Poison
freq          22
Name: track_name, dtype: object
--------------------
Column: track_artist
count             32828
unique            10692
top       Martin Garrix
freq                161
Name: track_artist, dtype: object
--------------------
Column: track_album_id
count                      32833
unique                     22545
top       5L1xcowSxwzFUSJzvyMp48
freq                          42
Name: track_album_id, dtype: object
--------------------
Column: track_album_name
count             32828
unique            19743
top       Greatest Hits
freq                139
Name: track_album_name, dtype: object
--------------------
Column: track_album_release_date
count          32833
unique          4530
top       2020-01-10


In [ ]:
df.describe()

track_popularity  danceability        energy           key  \
count      32833.000000  32833.000000  32833.000000  32833.000000   
mean          42.477081      0.654850      0.698619      5.374471   
std           24.984074      0.145085      0.180910      3.611657   
min            0.000000      0.000000      0.000175      0.000000   
25%           24.000000      0.563000      0.581000      2.000000   
50%           45.000000      0.672000      0.721000      6.000000   
75%           62.000000      0.761000      0.840000      9.000000   
max          100.000000      0.983000      1.000000     11.000000   

           loudness          mode   speechiness  acousticness  \
count  32833.000000  32833.000000  32833.000000  32833.000000   
mean      -6.719499      0.565711      0.107068      0.175334   
std        2.988436      0.495671      0.101314      0.219633   
min      -46.448000      0.000000      0.000000      0.000000   
25%       -8.171000      0.000000      0.041000      0.015100   
50%       -6.166000      1.000000      0.062500      0.080400   
75%       -4.645000      1.000000      0.132000      0.255000   
max        1.275000      1.000000      0.918000      0.994000   

       instrumentalness      liveness       valence         tempo  \
count      32833.000000  32833.000000  32833.000000  32833.000000   
mean           0.084747      0.190176      0.510561    120.881132   
std            0.224230      0.154317      0.233146     26.903624   
min            0.000000      0.000000      0.000000      0.000000   
25%            0.000000      0.092700      0.331000     99.960000   
50%            0.000016      0.127000      0.512000    121.984000   
75%            0.004830      0.248000      0.693000    133.918000   
max            0.994000      0.996000      0.991000    239.440000   

         duration_ms  
count   32833.000000  
mean   225799.811622  
std     59834.006182  
min      4000.000000  
25%    187819.000000  
50%    216000.000000  
75%    253585.000000  
max    517810.000000

# Problem Question 1

Bagaimana distribusi rata-rata popularitas lagu di setiap genre? Genre mana yang memiliki rata-rata popularitas tertinggi?

In [ ]:
# prompt: buatkan visualisasi berupa grafik untuk track_popularity on playlist_genre

# Calculate total track popularity per playlist genre
genre_popularity = df.groupby('playlist_genre')['track_popularity'].mean().reset_index()

# Create the visualization
fig = px.bar(genre_popularity, x='playlist_genre', y='track_popularity',
             title='Average Track Popularity per Playlist Genre',
             labels={'playlist_genre': 'Playlist Genre', 'track_popularity': 'Average Track Popularity'},
             color='playlist_genre')
fig.update_layout(xaxis_title='Playlist Genre', yaxis_title='Average Track Popularity')
fig.show()

1. **Distribusi Rata-rata Popularitas Lagu**:
   - **EDM** memiliki rata-rata popularitas sebesar **34.83**, menjadi genre dengan rata-rata popularitas paling rendah.
   - **Latin** memiliki rata-rata popularitas **47.03**, menunjukkan popularitas yang cukup tinggi.
   - **Pop** memiliki rata-rata popularitas tertinggi, yaitu **47.74**, menjadikannya genre yang paling populer.
   - **R&B** memiliki rata-rata popularitas sebesar **41.22**, berada di tingkat menengah.
   - **Rap** memiliki rata-rata popularitas **43.22**, sedikit lebih tinggi dibandingkan R&B.
   - **Rock** memiliki rata-rata popularitas sebesar **41.73**, mendekati nilai R&B.

2. **Genre dengan Rata-rata Popularitas Tertinggi**:
   - Genre **Pop** menempati posisi teratas dengan rata-rata popularitas **47.74**, sedikit lebih tinggi dari Latin (**47.03**), yang berada di posisi kedua.

3. **Genre dengan Rata-rata Popularitas Terendah**:
   - Genre **EDM** memiliki rata-rata popularitas terendah, yaitu **34.83**, menunjukkan bahwa genre ini kurang diminati dibandingkan genre lainnya.

# Problem Question 2

Siapa penyanyi dengan jumlah lagu terbanyak?

In [ ]:
# prompt: buatkan visualisasi list of song recorded by each singer top 20

import pandas as pd
import plotly.express as px

# Assuming your dataframe is named 'df'
# Calculate the top 20 singers by number of songs
top_singers = df.groupby('track_artist')['track_name'].count().sort_values(ascending=False).head(20)

# Create the visualization
fig = px.bar(
    top_singers,
    x=top_singers.index,
    y=top_singers.values,
    title="Top 20 Singers by Number of Songs Recorded",
    labels={"x": "Singer", "y": "Number of Songs"},
    color=top_singers.values
)
fig.update_layout(xaxis_title='Singer', yaxis_title='Number of Songs')
fig.show()

### 1. **Distribusi Jumlah Lagu yang Direkam oleh Penyanyi**:
   - **Martin Garrix** berada di posisi pertama dengan jumlah lagu terbanyak, yaitu **161 lagu**. Ini menunjukkan bahwa Martin Garrix adalah salah satu artis yang paling produktif dalam daftar ini.
   - **Queen** menempati posisi kedua dengan **136 lagu**, diikuti oleh **The Chainsmokers** dengan **123 lagu**.
   - Penyanyi lain seperti **David Guetta (110 lagu)**, **Don Omar (102 lagu)**, dan **Drake (100 lagu)** juga menunjukkan jumlah lagu yang signifikan, dengan rata-rata di atas 100 lagu.

### 2. **Artis dengan Jumlah Lagu Lebih Rendah pada Top 20 Singers**:
   - Penyanyi seperti **Hardwell (84 lagu)**, **Kygo (83 lagu)**, dan **Guns N' Roses (79 lagu)** memiliki jumlah lagu yang lebih sedikit dibandingkan dengan posisi teratas.
   - **Logic**, **Daddy Yankee**, dan **Tiësto** masing-masing memiliki jumlah lagu yang sama, yaitu **66 lagu**, menempati posisi terbawah dalam daftar ini.

### 3. **Penyanyi dengan Popularitas Menengah pada Top 20 Singers**:
   - Penyanyi seperti **The Weeknd (76 lagu)**, **Avicii (76 lagu)**, dan **J Balvin (72 lagu)** menunjukkan jumlah lagu yang cukup konsisten di kelas menengah.


### 4. **Rata-rata jumlah lagu yang direkam oleh top 20 penyanyi adalah 89.55 lagu**.

# Problem Question 3

Bagaimana rata-rata popularitas lagu berdasarkan genre playlist?

In [ ]:
# prompt: buatkan top 10 playlist sub genre

import pandas as pd
import plotly.express as px

# Assuming your dataframe is named 'df'
# Calculate the top 10 sub genres by number of songs
top_subgenres = df.groupby('playlist_genre')['track_name'].count().sort_values(ascending=False).head(10)

# Create the visualization
fig = px.bar(
    top_subgenres,
    x=top_subgenres.index,
    y=top_subgenres.values,
    title="Top Genres by Number of Songs",
    labels={"x": "Genre", "y": "Number of Songs"},
    color=top_subgenres.values
)
fig.update_layout(xaxis_title='Genre', yaxis_title='Number of Songs')
fig.show()

1. **Distribusi Jumlah lagu pada genre**


*   **edm** memiliki jumlah lagu sebanyak 6.043 lagu, menjadi genre yang memiliki jumlah lagu terbanyak
*   **rap** memiliki jumlah lagu sebanyak 5.743 lagu, menjadi genre memiliki lagu terbanyak setelah genre edm
*   **pop** memiliki jumlah lagu sebanyak 5.507 lagu, menjadi genre yang di tingkat menengah
*   **r&b** memiliki jumlah lagu sebanyak 5.431 lagu , genre ini niali nya mendekati jumlah dari genre pop
*   **latin** memiliki jumlah lagu sebanyak 5.153 lagu, menjadi genre yang memiliki jumlah lagu sedikit setelah genre rock
*   **rock**  memiliki jumlah lagu sebanyak 4.951 lagu, menjadi genre yang memiliki jumlah lagu sedikit

2. urutan genre

*   Urutan Genre Berdasarkan Jumlah Lagu:

  **EDM** > **Rap** > **Pop** > **R&B** > **Latin** > **Rock**.

*   Genre EDM sangat menonjol dibandingkan genre lainnya, sedangkan genre Rock memiliki jumlah lagu paling sedikit dalam daftar.

# Problem Quesiton 4

Apakah penyanyi tertentu cenderung memiliki popularitas lagu yang lebih tinggi dibandingkan penyanyi lainnya?

In [ ]:
# prompt: buatkan visualisasi dengan grafik line top 10 popular singer

import plotly.express as px

# Calculate the top 10 singers by track popularity
top_10_singers = df.groupby('track_artist')['track_popularity'].mean().sort_values(ascending=False).head(10)

# Create the line plot
fig = px.line(
    top_10_singers,
    x=top_10_singers.index,
    y=top_10_singers.values,
    title='Top 10 Most Popular Singers',
    labels={'x': 'Singer', 'y': 'Average Track Popularity'},
    markers=True  # Add markers to the line plot
)

fig.update_layout(xaxis_title='Singer', yaxis_title='Average Track Popularity')
fig.update_xaxes(tickangle=45)  # Rotate x-axis labels for better readability
fig.show()

1. Pada setiap musisi memiliki tingkat popularitas masing - masing, hal tersebut di tunjukan dengan grafik pada top 10 artis.
2. List populer musisi

*   Trevor daniel dengan rata rata kepopuleran lagu yaitu 96
*   Y2K dengan rata rata kepopuleran lagu yaitu 91
*   Don Toliver dengan rata rata kepopuleran lagu yaitu 90.7
*   Roddy Ricch dengan rata rata kepopuleran lagu yaitu 88.2
*   DaBaby dengan rata rata kepopuleran lagu yaitu 87.9
*   Kina dengan rata rata kepopuleran lagu yaitu 85.4
*   JACKBOYS dengan rata rata kepopuleran lagu yaitu 85.1
* YNW Melly dengan rata rata kepopuleran lagu yaitu 84.6
* Tainy, dan Dada Boladao dengan rata rata kepopuleran lagu yaitu 84

# Problem Question 5

Apakah ada pola tertentu yang menunjukkan hubungan antara tahun rilis album dan popularitas lagunya?

In [ ]:
import pandas as pd
import plotly.express as px

# Assuming your dataframe is named 'df' and it has columns 'track_popularity' and 'track_album_release_date'

# Convert 'track_album_release_date' to datetime objects with mixed format handling
df['track_album_release_date'] = pd.to_datetime(df['track_album_release_date'], format='mixed', errors='coerce')

# Drop rows with invalid dates
df = df.dropna(subset=['track_album_release_date'])

# Extract the year from the release date
df['release_year'] = df['track_album_release_date'].dt.year

# Group by release year and calculate the average popularity
avg_popularity_per_year = df.groupby('release_year')['track_popularity'].mean().reset_index()

# Create the line plot
fig = px.line(avg_popularity_per_year, x='release_year', y='track_popularity',
              title='Trend of Average Track Popularity Over Album Release Years',
              labels={'release_year': 'Album Release Year', 'track_popularity': 'Average Track Popularity'})

fig.update_layout(xaxis_title='Album Release Year', yaxis_title='Average Track Popularity')
fig.show()


### 1. **Pola Popularitas Lagu Berdasarkan Tahun Rilis**:
   - **Era Awal (1957-1970)**:
     - Popularitas rata-rata cenderung fluktuatif. Pada tahun-tahun awal, seperti 1957 (**30.0**) dan 1960 (**16.0**), popularitas cenderung lebih rendah.
     - Memasuki pertengahan hingga akhir 1960-an, popularitas meningkat, dengan puncaknya pada 1968 (**58.22**).

   - **Era 1970-an (1971-1980)**:
     - Popularitas rata-rata lebih stabil di kisaran 40-50, meskipun ada penurunan pada tahun tertentu, seperti 1974 (**40.39**).
     - Tahun 1980 menunjukkan peningkatan dengan popularitas rata-rata **52.25**, lebih tinggi dari dekade sebelumnya.

   - **Era 1980-an hingga 1990-an**:
     - Popularitas rata-rata cenderung mengalami penurunan bertahap dari awal 1980-an (sekitar **52.25** di tahun 1980) hingga akhir 1990-an (**41.79** pada 1999).
     - Secara umum, tren menunjukkan penurunan bertahap dalam popularitas.

   - **Era 2000-an hingga 2010-an**:
     - Popularitas rata-rata terus menurun secara signifikan, dengan titik terendah pada tahun 2008 (**27.26**) dan 2009 (**29.05**).
     - Setelah 2010, terjadi sedikit peningkatan hingga mencapai **41.76** pada tahun 2017.

   - **Era Modern (2018-2020)**:
     - Popularitas rata-rata meningkat pesat pada tahun-tahun terakhir, dari **45.88** (2018) hingga **51.40** (2019), sebelum sedikit menurun menjadi **46.46** pada tahun 2020.

### 2. **Kesimpulan Pola**:
   - Popularitas lagu cenderung tinggi pada akhir 1960-an hingga awal 1980-an, menunjukkan era di mana lagu-lagu klasik mendominasi.
   - Popularitas rata-rata menurun secara bertahap dari pertengahan 1980-an hingga akhir 2000-an, kemungkinan karena perubahan tren musik atau data yang lebih sedikit dari tahun-tahun tersebut.
   - Era modern (2018-2020) menunjukkan peningkatan popularitas rata-rata, yang dapat dikaitkan dengan peningkatan aksesibilitas musik melalui platform digital dan fokus pada lagu-lagu populer dalam dataset.

### 3. **Hubungan Tahun Rilis dengan Popularitas**:
   - Lagu-lagu yang dirilis pada tahun yang lebih baru (2018-2020) memiliki popularitas yang relatif lebih tinggi dibandingkan dengan dekade sebelumnya, menunjukkan bahwa musik modern lebih banyak didengarkan dan diakui.
   - Lagu-lagu klasik (1960-an hingga awal 1980-an) juga memiliki popularitas tinggi, kemungkinan karena dampaknya yang besar di era tersebut dan apresiasi lintas generasi.

# Conclusion

1. **Popularitas Genre Musik**:  
   - Genre **Pop** memiliki rata-rata popularitas tertinggi (**47.74**), diikuti oleh **Latin** (**47.03**) dan **Rap** (**43.22**). Sebaliknya, genre **EDM** memiliki rata-rata popularitas terendah (**34.83**), menunjukkan daya tarik yang lebih rendah dibandingkan genre lainnya.  
   - Urutan berdasarkan jumlah lagu menunjukkan **EDM** sebagai genre dengan lagu terbanyak (6.043 lagu), sedangkan **Rock** memiliki jumlah lagu paling sedikit (4.951 lagu).  

2. **Produktivitas Penyanyi**:  
   - **Martin Garrix** menjadi artis dengan jumlah lagu terbanyak (**161 lagu**), diikuti oleh **Queen** (**136 lagu**) dan **The Chainsmokers** (**123 lagu**). Penyanyi dengan jumlah lagu lebih sedikit seperti **Hardwell** (**84 lagu**) dan **Kygo** (**83 lagu**) masih berada di posisi teratas namun kurang produktif dibandingkan pemimpin daftar.  

3. **Tren Popularitas Berdasarkan Tahun Rilis**:  
   - Popularitas lagu mencapai puncak pada akhir 1960-an hingga awal 1980-an, kemudian menurun hingga akhir 2000-an. Era modern (2018-2020) menunjukkan peningkatan kembali, dengan popularitas rata-rata tertinggi pada 2019 (**51.40**). Lagu-lagu klasik dari 1960-an hingga awal 1980-an tetap dihargai lintas generasi, sementara musik modern mendapatkan perhatian lebih besar karena aksesibilitas digital.  

4. **Penyanyi dengan Popularitas Tinggi**:  
   - **Trevor Daniel** memiliki rata-rata popularitas lagu tertinggi (**96**), diikuti oleh **Y2K** (**91**) dan **Don Toliver** (**90.7**). Penyanyi seperti **Roddy Ricch**, **DaBaby**, dan **Kina** juga menunjukkan rata-rata popularitas yang konsisten di atas **85**, mencerminkan daya tarik kuat mereka dalam industri musik modern.  